// @strapi/plugin-users-permissions/dist/admin/chunks/index-DAclA-0k.js
'use strict';

var name$1 = "@strapi/plugin-users-permissions";
var strapi = {
    displayName: "Roles & Permissions",
    name: "users-permissions",
    description: "Protect your API with a full authentication process based on JWT. This plugin comes also with an ACL strategy that allows you to manage the permissions between the groups of users.",
    required: true,
    kind: "plugin"
};

const PERMISSIONS = {
    // Roles
    accessRoles: [
        {
            action: 'plugin::users-permissions.roles.create',
            subject: null
        },
        {
            action: 'plugin::users-permissions.roles.read',
            subject: null
        }
    ],
    createRole: [
        {
            action: 'plugin::users-permissions.roles.create',
            subject: null
        }
    ],
    deleteRole: [
        {
            action: 'plugin::users-permissions.roles.delete',
            subject: null
        }
    ],
    readRoles: [
        {
            action: 'plugin::users-permissions.roles.read',
            subject: null
        }
    ],
    updateRole: [
        {
            action: 'plugin::users-permissions.roles.update',
            subject: null
        }
    ],
    // AdvancedSettings
    readAdvancedSettings: [
        {
            action: 'plugin::users-permissions.advanced-settings.read',
            subject: null
        }
    ],
    updateAdvancedSettings: [
        {
            action: 'plugin::users-permissions.advanced-settings.update',
            subject: null
        }
    ],
    // Emails
    readEmailTemplates: [
        {
            action: 'plugin::users-permissions.email-templates.read',
            subject: null
        }
    ],
    updateEmailTemplates: [
        {
            action: 'plugin::users-permissions.email-templates.update',
            subject: null
        }
    ],
    // Providers
    readProviders: [
        {
            action: 'plugin::users-permissions.providers.read',
            subject: null
        }
    ],
    updateProviders: [
        {
            action: 'plugin::users-permissions.providers.update',
            subject: null
        }
    ]
};

const pluginId = name$1.replace(/^@strapi\/plugin-/i, '');

const getTrad = (id)=>`${pluginId}.${id}`;

const prefixPluginTranslations = (trad, pluginId)=>{
    return Object.keys(trad).reduce((acc, current)=>{
        acc[`${pluginId}.${current}`] = trad[current];
        return acc;
    }, {});
};

function __variableDynamicImportRuntime4__(path) {
  switch (path) {
    case './translations/ar.json': return Promise.resolve().then(function () { return require('./ar-BJwjobLp.js'); });
    case './translations/cs.json': return Promise.resolve().then(function () { return require('./cs-uS_SIEo8.js'); });
    case './translations/de.json': return Promise.resolve().then(function () { return require('./de-7MVMrqqI.js'); });
    case './translations/dk.json': return Promise.resolve().then(function () { return require('./dk-DwCLGmy9.js'); });
    case './translations/en.json': return Promise.resolve().then(function () { return require('./en-DwQjkHi_.js'); });
    case './translations/es.json': return Promise.resolve().then(function () { return require('./es-BOJOedG5.js'); });
    case './translations/fr.json': return Promise.resolve().then(function () { return require('./fr-BDNWCNs0.js'); });
    case './translations/id.json': return Promise.resolve().then(function () { return require('./id-UqUPykHZ.js'); });
    case './translations/it.json': return Promise.resolve().then(function () { return require('./it-D5VuyoLU.js'); });
    case './translations/ja.json': return Promise.resolve().then(function () { return require('./ja-MpqVsCgs.js'); });
    case './translations/ko.json': return Promise.resolve().then(function () { return require('./ko-Bm-grPSc.js'); });
    case './translations/ms.json': return Promise.resolve().then(function () { return require('./ms-D7eyBD5H.js'); });
    case './translations/nl.json': return Promise.resolve().then(function () { return require('./nl-BIOwAQtI.js'); });
    case './translations/pl.json': return Promise.resolve().then(function () { return require('./pl-D5BeNrg_.js'); });
    case './translations/pt-BR.json': return Promise.resolve().then(function () { return require('./pt-BR-DxPBzQGx.js'); });
    case './translations/pt.json': return Promise.resolve().then(function () { return require('./pt-kkCwzNvH.js'); });
    case './translations/ru.json': return Promise.resolve().then(function () { return require('./ru-BQ0gHmp3.js'); });
    case './translations/sk.json': return Promise.resolve().then(function () { return require('./sk-nVwAPdYC.js'); });
    case './translations/sv.json': return Promise.resolve().then(function () { return require('./sv-BDfk2A-F.js'); });
    case './translations/th.json': return Promise.resolve().then(function () { return require('./th-BtTtpHe2.js'); });
    case './translations/tr.json': return Promise.resolve().then(function () { return require('./tr-Di-Nf7cT.js'); });
    case './translations/uk.json': return Promise.resolve().then(function () { return require('./uk-CggQOx1l.js'); });
    case './translations/vi.json': return Promise.resolve().then(function () { return require('./vi-D9cCsHsU.js'); });
    case './translations/zh-Hans.json': return Promise.resolve().then(function () { return require('./zh-Hans-CKqQbpsM.js'); });
    case './translations/zh.json': return Promise.resolve().then(function () { return require('./zh-BzWgJEzz.js'); });
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }
const name = strapi.name;
var index = {
    register (app) {
        // Create the plugin's settings section
        app.createSettingSection({
            id: 'users-permissions',
            intlLabel: {
                id: getTrad('Settings.section-label'),
                defaultMessage: 'Users & Permissions plugin'
            }
        }, [
            {
                intlLabel: {
                    id: 'global.roles',
                    defaultMessage: 'Roles'
                },
                id: 'roles',
                to: `users-permissions/roles`,
                Component: ()=>Promise.resolve().then(function () { return require('./index-D01zzG9y.js'); }),
                permissions: PERMISSIONS.accessRoles
            },
            {
                intlLabel: {
                    id: getTrad('HeaderNav.link.providers'),
                    defaultMessage: 'Providers'
                },
                id: 'providers',
                to: `users-permissions/providers`,
                Component: ()=>Promise.resolve().then(function () { return require('./index-BtYUb_br.js'); }),
                permissions: PERMISSIONS.readProviders
            },
            {
                intlLabel: {
                    id: getTrad('HeaderNav.link.emailTemplates'),
                    defaultMessage: 'Email templates'
                },
                id: 'email-templates',
                to: `users-permissions/email-templates`,
                Component: ()=>Promise.resolve().then(function () { return require('./index-BPiDUOGt.js'); }).then((mod)=>({
                            default: mod.ProtectedEmailTemplatesPage
                        })),
                permissions: PERMISSIONS.readEmailTemplates
            },
            {
                intlLabel: {
                    id: getTrad('HeaderNav.link.advancedSettings'),
                    defaultMessage: 'Advanced Settings'
                },
                id: 'advanced-settings',
                to: `users-permissions/advanced-settings`,
                Component: ()=>Promise.resolve().then(function () { return require('./index-CkYplz_3.js'); }).then((mod)=>({
                            default: mod.ProtectedAdvancedSettingsPage
                        })),
                permissions: PERMISSIONS.readAdvancedSettings
            }
        ]);
        app.registerPlugin({
            id: 'users-permissions',
            name
        });
    },
    bootstrap () {},
    async registerTrads ({ locales }) {
        const importedTrads = await Promise.all(locales.map((locale)=>{
            return __variableDynamicImportRuntime4__(`./translations/${locale}.json`).then(({ default: data })=>{
                return {
                    data: prefixPluginTranslations(data, 'users-permissions'),
                    locale
                };
            }).catch(()=>{
                return {
                    data: {},
                    locale
                };
            });
        }));
        return Promise.resolve(importedTrads);
    }
};

exports.PERMISSIONS = PERMISSIONS;
exports.getTrad = getTrad;
exports.index = index;
