// @strapi/admin/dist/admin/chunks/PurchaseAuditLogs-BTAi887h.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var symbols = require('@strapi/icons/symbols');
var reactIntl = require('react-intl');
var index = require('./index-D7OmC33R.js');
require('react-dom/client');
require('react');
require('invariant');
require('lodash/isFunction');
require('lodash/merge');
require('lodash/pick');
require('react-router-dom');
require('immer');
require('./Theme-CFGbtvHi.js');
require('@radix-ui/react-context');
require('qs');
require('./admin-udBiOv2o.js');
require('lodash/trimEnd');
require('@reduxjs/toolkit/query/react');
require('lodash/fp/pipe');
require('lodash/clone');
require('lodash/toPath');
require('lodash/isEqual');
require('axios');
require('lodash/get');
require('lodash/set');
require('use-context-selector');
require('lodash/defaultsDeep');
require('react-redux');
require('styled-components');
require('@reduxjs/toolkit');
require('react-query');
require('./useEnterprise-IHd3AKd2.js');
require('lodash/camelCase');
require('yup');
require('fractional-indexing');
require('lodash/omit');
require('formik');
require('lodash/throttle');

const PurchaseAuditLogs = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    return /*#__PURE__*/ jsxRuntime.jsx(index.Layouts.Root, {
        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Main, {
            children: [
                /*#__PURE__*/ jsxRuntime.jsx(index.Layouts.Header, {
                    title: formatMessage({
                        id: 'global.auditLogs',
                        defaultMessage: 'Audit Logs'
                    }),
                    subtitle: formatMessage({
                        id: 'Settings.permissions.auditLogs.listview.header.subtitle',
                        defaultMessage: 'Logs of all the activities that happened in your environment'
                    })
                }),
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                    paddingLeft: 10,
                    paddingRight: 10,
                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.EmptyStateLayout, {
                        icon: /*#__PURE__*/ jsxRuntime.jsx(symbols.EmptyPermissions, {
                            width: "16rem"
                        }),
                        content: formatMessage({
                            id: 'Settings.permissions.auditLogs.not-available',
                            defaultMessage: 'Audit Logs is only available as part of a paid plan. Upgrade to get a searchable and filterable display of all activities.'
                        }),
                        action: /*#__PURE__*/ jsxRuntime.jsx(designSystem.LinkButton, {
                            variant: "default",
                            endIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.ExternalLink, {}),
                            href: "https://strapi.io/features/audit-logs?utm_campaign=In-Product-CTA&utm_source=Audit-Logs",
                            isExternal: true,
                            target: "_blank",
                            children: formatMessage({
                                id: 'global.learn-more',
                                defaultMessage: 'Learn more'
                            })
                        })
                    })
                })
            ]
        })
    });
};

exports.PurchaseAuditLogs = PurchaseAuditLogs;
