// @strapi/admin/dist/admin/chunks/PurchaseAuditLogs-McvZESGI.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { Main, Box, EmptyStateLayout, LinkButton } from '@strapi/design-system';
import { ExternalLink } from '@strapi/icons';
import { EmptyPermissions } from '@strapi/icons/symbols';
import { useIntl } from 'react-intl';
import { b as Layouts } from './index-BUGouNML.mjs';
import 'react-dom/client';
import 'react';
import 'invariant';
import 'lodash/isFunction';
import 'lodash/merge';
import 'lodash/pick';
import 'react-router-dom';
import 'immer';
import './Theme-CDaBfQ-7.mjs';
import '@radix-ui/react-context';
import 'qs';
import './admin-BBS7sDVu.mjs';
import 'lodash/trimEnd';
import '@reduxjs/toolkit/query/react';
import 'lodash/fp/pipe';
import 'lodash/clone';
import 'lodash/toPath';
import 'lodash/isEqual';
import 'axios';
import 'lodash/get';
import 'lodash/set';
import 'use-context-selector';
import 'lodash/defaultsDeep';
import 'react-redux';
import 'styled-components';
import '@reduxjs/toolkit';
import 'react-query';
import './useEnterprise-Dy8wMNV5.mjs';
import 'lodash/camelCase';
import 'yup';
import 'fractional-indexing';
import 'lodash/omit';
import 'formik';
import 'lodash/throttle';

const PurchaseAuditLogs = ()=>{
    const { formatMessage } = useIntl();
    return /*#__PURE__*/ jsx(Layouts.Root, {
        children: /*#__PURE__*/ jsxs(Main, {
            children: [
                /*#__PURE__*/ jsx(Layouts.Header, {
                    title: formatMessage({
                        id: 'global.auditLogs',
                        defaultMessage: 'Audit Logs'
                    }),
                    subtitle: formatMessage({
                        id: 'Settings.permissions.auditLogs.listview.header.subtitle',
                        defaultMessage: 'Logs of all the activities that happened in your environment'
                    })
                }),
                /*#__PURE__*/ jsx(Box, {
                    paddingLeft: 10,
                    paddingRight: 10,
                    children: /*#__PURE__*/ jsx(EmptyStateLayout, {
                        icon: /*#__PURE__*/ jsx(EmptyPermissions, {
                            width: "16rem"
                        }),
                        content: formatMessage({
                            id: 'Settings.permissions.auditLogs.not-available',
                            defaultMessage: 'Audit Logs is only available as part of a paid plan. Upgrade to get a searchable and filterable display of all activities.'
                        }),
                        action: /*#__PURE__*/ jsx(LinkButton, {
                            variant: "default",
                            endIcon: /*#__PURE__*/ jsx(ExternalLink, {}),
                            href: "https://strapi.io/features/audit-logs?utm_campaign=In-Product-CTA&utm_source=Audit-Logs",
                            isExternal: true,
                            target: "_blank",
                            children: formatMessage({
                                id: 'global.learn-more',
                                defaultMessage: 'Learn more'
                            })
                        })
                    })
                })
            ]
        })
    });
};

export { PurchaseAuditLogs };
