// @strapi/plugin-cloud/dist/admin/chunks/index-DBuRsSPK.js
'use strict';

var icons = require('@strapi/icons');
var react = require('react');

const pluginId = 'cloud';

const Initializer = ({ setPlugin })=>{
    const ref = react.useRef(setPlugin);
    react.useEffect(()=>{
        ref.current(pluginId);
    }, []);
    return null;
};

const prefixPluginTranslations = (trad, pluginId)=>{
    return Object.keys(trad).reduce((acc, current)=>{
        acc[`${pluginId}.${current}`] = trad[current];
        return acc;
    }, {});
};

function __variableDynamicImportRuntime1__(path) {
  switch (path) {
    case './translations/en.json': return Promise.resolve().then(function () { return require('./en-uuzupUSq.js'); });
    case './translations/fr.json': return Promise.resolve().then(function () { return require('./fr-QQbSxlps.js'); });
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }
const name = 'Deploy';
// eslint-disable-next-line import/no-default-export
var index = {
    register (app) {
        const { backendURL } = window.strapi;
        // Only add the plugin menu link and registering it if the project is on development (localhost).
        if (backendURL?.includes('localhost')) {
            app.addMenuLink({
                to: `plugins/${pluginId}`,
                icon: icons.Cloud,
                intlLabel: {
                    id: `${pluginId}.plugin.name`,
                    defaultMessage: name
                },
                Component: async ()=>{
                    const { App } = await Promise.resolve().then(function () { return require('./App-CwT9IBBa.js'); });
                    return App;
                }
            });
            const plugin = {
                id: pluginId,
                initializer: Initializer,
                isReady: false,
                name
            };
            app.registerPlugin(plugin);
        }
    },
    async registerTrads (app) {
        const { locales } = app;
        const importedTrads = await Promise.all(locales.map((locale)=>{
            return __variableDynamicImportRuntime1__(`./translations/${locale}.json`).then(({ default: data })=>{
                return {
                    data: prefixPluginTranslations(data, pluginId),
                    locale
                };
            }).catch(()=>{
                return {
                    data: {},
                    locale
                };
            });
        }));
        return Promise.resolve(importedTrads);
    }
};

exports.index = index;
exports.pluginId = pluginId;
